Add gtdbtk/classifywf module

abhi18av opened the pull request:
<!--
# nf-core/modules pull request

Many thanks for contributing to nf-core/modules!

Please fill in the appropriate checklist below (delete whatever is not relevant).
These are the most common things requested on pull requests (PRs).

Remember that PRs should be made against the master branch.

Learn more about contributing: [CONTRIBUTING.md](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
-->

## PR checklist

Closes #XXX <!-- If this PR fixes an issue, please link it here! -->

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [x] If necessary, include test data in your PR.
- [x] Remove all TODO statements.
- [x] Emit the `versions.yml` file.
- [x] Follow the naming conventions.
- [x] Follow the parameters requirements.
- [x] Follow the input/output options guidelines.
- [x] Add a resource `label`
- [x] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
    - [x] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd`
    - [x] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd`
    - [x] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd`

abhi18av commented on modules/gtdbtk/classifywf/main.nf:
```diff
@@ -0,0 +1,60 @@
+include { initOptions; saveFiles; getSoftwareName } from './functions'
+
+params.options = [:]
+options    = initOptions(params.options)
+
+process GTDBTK_CLASSIFY {
+    tag "${meta.assembler}-${meta.id}"
+
+    publishDir "${params.outdir}",
+        mode: params.publish_dir_mode,
+        saveAs: { filename -> saveFiles(filename:filename, options:params.options, publish_dir:getSoftwareName(task.process), meta:meta, publish_by_meta:['assembler', 'id']) }
```
As of now I have re-used the code available at https://github.com/nf-core/mag/blob/master/modules/local/gtdbtk_classify.nf.

abhi18av commented on modules/gtdbtk/classifywf/meta.yml:
```diff
@@ -0,0 +1,78 @@
+  - version:
+      type: file
+      description: File containing software version
+      pattern: "*.{version.txt}"
+  - summary:
+      type: file
+      description: A TSV summary file for the classification
+      pattern: "*.{summary.tsv}"
+  - tree:
+      type: file
+      description: NJ or UPGMA tree in Newick format produced from a multiple sequence alignment
+      pattern: "*.{classify.tree.gz}"
+  - markers:
+      type: file
+      description: A TSV summary file lineage markers used for the classification.
+      pattern: "*.{markers_summary.tsv}"
+  - msa:
+      type: file
+      description: Multiple sequence alignments file.
+      pattern: "*.{msa.fasta.gz}"
+  - user_msa:
+      type: file
+      description: Multiple sequence alignments file for the user-provided files.
+      pattern: "*.{user_msa.fasta.gz}"
+  - filtered:
+      type: file
+      description: A TSV summary of the filtered genomes which were used in the GTDB-tk classification.
+      pattern: "*.{filtered.tsv}"
+  - log:
+      type: file
+      description: GTDB-tk log file
+      pattern: "*.{log}"
+  - warnings:
+      type: file
+      description: GTDB-tk warnings log file
+      pattern: "*.{warnings.log}"
+  - failed:
+      type: file
+      description: A TSV summary of the genomes which GTDB-tk failed to classify.
+      pattern: "*.{failed_genomes.tsv}"
```
It'd be great if someone could validate these descriptions please.

abhi18av commented on tests/modules/gtdbtk/classifywf/main.nf:
```diff
@@ -0,0 +1,13 @@
+
+include { GTDBTK_CLASSIFYWF } from '../../../../modules/gtdbtk/classifywf/main.nf' addParams( options: [:] )
+
+workflow test_gtdbtk_classifywf {
+    
+    input = [ [ id:'test', single_end:false ], // meta map
+              file(params.test_data['sarscov2']['illumina']['test_paired_end_bam'], checkIfExists: true) ]
+
+    GTDBTK_CLASSIFYWF ( input )
+}
```
And this is the complicate part, I am not sure how to test this module since it's requires the presence of a sizeable taxonomic database on disk.

rpetit3 replied:
I think for this lets do stubs, and we can verify locally

abhi18av replied:
Cool, this makes sense 👍

abhi18av reviewed (commented):
I have added my comments for the reviewer - thanks for your time 🙏 

For some context: This module is targeted towards [Bactopia v2 ](https://github.com/bactopia/bactopia/issues/233#issuecomment-899593455) and hopefully it could be consumed by `nf-core/mag` as well.

abhi18av commented:
I've added the stubs for generting a database, but the tests are failing with an odd error 🤔 

I am not sure which format exactly the ci-tests are expecting the `test.yml` file.

```
    def test_ensure_valid_version_yml(workflow_dir):
        workflow_dir = Path(workflow_dir)
        software_name = workflow_dir.name.split("_")[0].lower()
        try:
            versions_yml_file = workflow_dir / f"output/{software_name}/versions.yml"
            versions_yml = versions_yml_file.read_text()
        except FileNotFoundError:
>           raise AssertionError(
                dedent(
                    f"""\
                    `versions.yml` not found in the output directory.
                    Expected path: `{versions_yml_file}`
    
                    This can have multiple reasons:
                    * The test-workflow failed before a `versions.yml` could be generated.
                    * The workflow name in `test.yml` does not start with the tool name.
                    """
                )
            )
E           AssertionError: `versions.yml` not found in the output directory.
E           Expected path: `/home/runner/pytest_workflow_f71wgx8f/gtdbtk_classifywf/output/gtdbtk/versions.yml`
E           
E           This can have multiple reasons:
E           * The test-workflow failed before a `versions.yml` could be generated.
E           * The workflow name in `test.yml` does not start with the tool name.

tests/test_versions_yml.py:29: AssertionError

```

rpetit3 commented:
SUper close, it'll need to be updated for the new changes in software versions

https://github.com/nf-core/modules/blob/master/modules/abacas/main.nf#L42-L45

abhi18av commented:
Hmm, this is an interesting change - thanks @rpetit3 , I need to hand around the `#modules` channel more :) 

However, for `gtdbtk` this is a bit problematic, since `gtdktk --version` doesn't work without the presence of the database and throws the following error 

```yaml
GTDBTK_CLASSIFYWF:
    gtdbtk: ================================================================================ ERROR ________________________________________________________________________________ The 'GTDBTK_DATA_PATH' environment variable is not defined. Please set this variable to your reference data package. https://github.com/Ecogenomics/GTDBTk#installation ================================================================================
```

... which not being a formatted YAML, raises issue with the `yaml` file validation 😶 

```
E           AssertionError: Version number for gtdbtk must start with a number. 
E           assert None
E            +  where None = <function match at 0x7f86fb3bfa60>('^\\d+.*', '================================================================================ ERROR ______________________________...b.com/Ecogenomics/GTDBTk#installation ================================================================================')
E            +    where <function match at 0x7f86fb3bfa60> = re.match
E            +    and   '================================================================================ ERROR ______________________________...b.com/Ecogenomics/GTDBTk#installation ================================================================================' = str('================================================================================ ERROR ______________________________...b.com/Ecogenomics/GTDBTk#installation ================================================================================')
```

The plot thickens 🕵️

abhi18av commented:
Hmm, I'm a bit confused here. Now that I've fixed the problem while printing the version for `gtdb` in the `stubs`, the `nf-core linting` throws the following error and the `conda` build is failing 

I'm using `nf-core, version 2.1` - should I switch to `dev` based package to keep these from failing?

```
╭──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╮
│ [!] 2 Tests Failed                                                                                                                   │
╰──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
╭──────────────────────────────────────────┬────────────────────────────────────────┬──────────────────────────────────────────────────╮
│ Module name                              │ File path                              │ Test message                                     │
├──────────────────────────────────────────┼────────────────────────────────────────┼──────────────────────────────────────────────────┤
│ gtdbtk/classifywf                        │ modules/gtdbtk/classifywf/main.nf      │ 'publish_by_meta:['id']' unspecified in saveAs   │
│                                          │                                        │ function                                         │
│ gtdbtk/classifywf                        │ modules/gtdbtk/classifywf/functions.nf │ New version of functions.nf available            │
╰──────────────────────────────────────────┴────────────────────────────────────────┴──────────────────────────────────────────────────╯
```

drpatelh commented on modules/gtdbtk/classifywf/main.nf:
```diff
@@ -0,0 +1,86 @@
+
+    publishDir "${params.outdir}",
+        mode: params.publish_dir_mode,
+        saveAs: { filename -> saveFiles(filename:filename, options:params.options, publish_dir:getSoftwareName(task.process), meta:meta, publish_by_meta:['assembler', 'id']) }
```
```suggestion
        saveAs: { filename -> saveFiles(filename:filename, options:params.options, publish_dir:getSoftwareName(task.process), meta:meta, publish_by_meta:['id']) }
```

abhi18av replied:
Ah, I see. Thanks for pointing this out - applying now.

Then, I presume that only the `id` field is considered during the publishing process?

drpatelh commented:
Weird about the `functions.nf` 🤔 Looks ok but maybe there is a `diff` there I am not seeing. 

Could you try to replace the one on the PR with [this one](https://github.com/nf-core/tools/blob/dev/nf_core/module-template/modules/functions.nf).

And apply the suggested change [above](https://github.com/nf-core/modules/pull/765#discussion_r721142056)?

abhi18av commented:
> Could you try to replace the one on the PR with this one.

Done 👍 


> And apply the suggested change above?

Done 👍 

The problem with `conda` still persists  🤔

rpetit3 commented on modules/gtdbtk/classifywf/meta.yml:
```diff
@@ -0,0 +1,78 @@
+      pattern: "*.{user_msa.fasta.gz}"
+  - filtered:
+      type: file
+      description: A TSV summary of the filtered genomes which were used in the GTDB-tk classification.
```
`A list of genomes with an insufficient number of amino acids in MSA.`

https://ecogenomics.github.io/GTDBTk/files/filtered.tsv.html

abhi18av replied:
I've update the description ✅

abhi18av replied:
Also good catch regarding the gtdb module being sourced from `conda-forge` - thanks Robert!

drpatelh commented on modules/gtdbtk/classifywf/main.nf:
```diff
@@ -0,0 +1,86 @@
+include { initOptions; saveFiles; getSoftwareName; getProcessName } from './functions'
+
+params.options = [:]
+options    = initOptions(params.options)
+
```
```suggestion
params.options = [:]
options        = initOptions(params.options)
```

drpatelh commented on modules/gtdbtk/classifywf/main.nf:
```diff
@@ -0,0 +1,86 @@
+        mkdir pplacer_tmp
+    fi
+
+    gtdbtk classify_wf $options.args \
+                    --genome_dir bins \
+                    --prefix "gtdbtk.${meta.assembler}-${meta.id}" \
+                    --out_dir "\${PWD}" \
+                    --cpus ${task.cpus} \
+                    --pplacer_cpus ${params.gtdbtk_pplacer_cpus} \
+                    ${pplacer_scratch} \
+                    --min_perc_aa ${params.gtdbtk_min_perc_aa} \
+                    --min_af ${params.gtdbtk_min_af}
```
```suggestion
    gtdbtk classify_wf \\
		  $options.args \\
		  --genome_dir bins \\
		  --prefix "gtdbtk.${meta.assembler}-${meta.id}" \\
		  --out_dir "\${PWD}" \\
		  --cpus ${task.cpus} \\
		  --pplacer_cpus $params.gtdbtk_pplacer_cpus \\
		  $pplacer_scratch \\
		  --min_perc_aa $params.gtdbtk_min_perc_aa \\
		  --min_af $params.gtdbtk_min_af
```

drpatelh commented on tests/modules/gtdbtk/classifywf/main.nf:
```diff
@@ -0,0 +1,32 @@
+
+    STUB_GTDBTK_DATABASE()
+
+    input = [ [ id:'test', single_end:false, assembler:'SPADES' ],
+              [ file(params.test_data['sarscov2']['genome']['genome_fasta'], checkIfExists: true),
+          	file(params.test_data['sarscov2']['illumina']['contigs_fasta'], checkIfExists: true),
+                file(params.test_data['sarscov2']['illumina']['scaffolds_fasta'], checkIfExists: true)]]
+
+
+
+    GTDBTK_CLASSIFYWF( input, STUB_GTDBTK_DATABASE.out.database )
```
```suggestion
    input = [ 
        [ id:'test', single_end:false, assembler:'SPADES' ],
        [ 
            file(params.test_data['sarscov2']['genome']['genome_fasta'], checkIfExists: true),
            file(params.test_data['sarscov2']['illumina']['contigs_fasta'], checkIfExists: true),
            file(params.test_data['sarscov2']['illumina']['scaffolds_fasta'], checkIfExists: true)
        ]
    ]
	
    GTDBTK_CLASSIFYWF ( input, STUB_GTDBTK_DATABASE.out.database )
```

drpatelh commented:
Thanks guys! This is a good start and can be amended later if required.

drpatelh commented on modules/gtdbtk/classifywf/meta.yml:
```diff
@@ -0,0 +1,78 @@
+        e.g. [ id:'test', single_end:false ]
+  - versions:
+      type: file
+      description: File containing software version
```
```suggestion
      description: File containing software versions
```
